src/KubernetesClient/Versioning/VersionConverter.cs

```diff
@@ -0,0 +1,179 @@
+        }
+        public static IMapper Mapper { get; private set; }
+        internal static MapperConfiguration MapperConfiguration { get; private set; }
+
+        /// <summary>
+        /// Two level lookup of model types by Kind and then Version
+        /// </summary>
+        internal static Dictionary<string,Dictionary<string,Type>> KindVersionsMap { get; private set; }
+        public static void UpdateMappingConfiguration(Action<IMapperConfigurationExpression> configuration)
+        {
+
+            MapperConfiguration = new MapperConfiguration(cfg =>
+            {
+                GetKubernetesMaps(cfg);
+                configuration(cfg);
+            });
+            Mapper = MapperConfiguration.CreateMapper();
+            KindVersionsMap = MapperConfiguration
+                .GetAllTypeMaps()
+                .SelectMany(x => new[] {x.Types.SourceType, x.Types.DestinationType})
+                .Where(x => x.GetCustomAttribute<KubernetesEntityAttribute>() != null)
+                .Select(x =>
+                {
+                    var attr = GetKubernetesEntityAttribute(x);
+                    return new {attr.Kind, attr.ApiVersion, Type = x};
+                })
+                .GroupBy(x => x.Kind)
+                .ToDictionary(x => x.Key, kindGroup => kindGroup
+                    .GroupBy(x => x.ApiVersion)
+                    .ToDictionary(x => x.Key,
+                        versionGroup => versionGroup.Select(x => x.Type).Distinct().Single())); // should only be one type for each Kind/Version combination
+        }
+
+        public static object ConvertToVersion(object source, string apiVersion)
+        {
+            var type = source.GetType();
+            var attr = GetKubernetesEntityAttribute(type);
+            if(!KindVersionsMap.TryGetValue(attr.Kind, out var kindVersions))
+                throw new InvalidOperationException($"Version converter does not have any registered types for Kind `{attr.Kind}`");
+            if(!kindVersions.TryGetValue(apiVersion, out var targetType) || !kindVersions.TryGetValue(attr.ApiVersion, out var sourceType) || MapperConfiguration.FindTypeMapFor(sourceType, targetType) == null)
+                throw new InvalidOperationException($"There is no conversion mapping registered for Kind `{attr.Kind}` from ApiVersion {attr.ApiVersion} to {apiVersion}");
+            return Mapper.Map(source, sourceType, targetType);
+        }
+
+        private static KubernetesEntityAttribute GetKubernetesEntityAttribute(Type type)
+        {
+            var attr = type.GetCustomAttribute<KubernetesEntityAttribute>();
+            if(attr == null)
+                throw new InvalidOperationException($"Type {type} does not have {nameof(KubernetesEntityAttribute)}");
+            return attr;
+        }
+
+        internal static void GetKubernetesMaps(IMapperConfigurationExpression cfg)
+        {
+            cfg.DisableConstructorMapping();
```

Comment:
generated?

Reply:
I generated it in LinqPad, but few of these maps had to be manually realigned. I guess I can put the autogeneration logic where there are no conflicts into the gen project and augment it with manual maps for where there are conflicts.